TASK: document and assert Node.nodeName behavior

mhsdesign opened the pull request:
related https://github.com/neos/neos-development-collection/issues/4311

<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

When looking into the php node readmodel it should be super clear that a nodename is guaranteed to be set if its tethered.

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
TASK: document and assert Node.nodeName behavior

related https://github.com/neos/neos-development-collection/issues/4311

mhsdesign commented:
Sadly phpstan doenst allow us to annotate this behavior  ...

https://phpstan.org/writing-php-code/narrowing-types#custom-type-checking-functions-and-methods

And also the feature like in `@return` doesnt work here https://phpstan.org/blog/phpstan-1-6-0-with-conditional-return-types:

```
/** @var ($this->classification is TETHRED ? string : null) */
```

playground:
https://phpstan.org/r/7c491f32-89df-4b4b-a218-00c2b1889933

mhsdesign commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php:
```diff
@@ -75,9 +75,23 @@ public function __construct(
+            || $this->nodeName instanceof NodeName,
+            'The NodeName must be set if the Node is tethered.'
+        );
+        assert(
```
remove check for nodeType

bwaidelich commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php:
```diff
@@ -75,9 +75,23 @@ public function __construct(
+        assert(
+            !$this->nodeType instanceof NodeType
+            || $this->nodeType->name->equals($this->nodeTypeName),
+            'The name of the NodeType must equal the NodeTypeName of the Node.'
```
in case you leave this in:
```suggestion
            'The name of the NodeType must be equal to the NodeTypeName of the Node.'
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php:
```diff
@@ -75,9 +75,23 @@ public function __construct(
         public ?NodeName $nodeName,
         public Timestamps $timestamps,
     ) {
+        assert(
+            $this->classification !== NodeAggregateClassification::CLASSIFICATION_TETHERED
+            || $this->nodeName instanceof NodeName,
+            'The NodeName must be set if the Node is tethered.'
+        );
```
One detail: [assert](https://www.php.net/manual/de/function.assert.php) is only evaluated if the corresponding PHP flag is set (and it's behavior is depending on those flags, too). So we should avoid it for productive code.
```suggestion
        if ($classification === NodeAggregateClassification::CLASSIFICATION_TETHERED && $nodeName === null) {
            throw new \InvalidArgumentException('The node name must be set if the node is tethered');
        }
```

bwaidelich reviewed (commented):
+1 for defensive style (even though people will probably not instantiate this class manually) because it makes invalid states impossible.
In the long run we will hopefully be able to remove `NodeName` and `NodeType` from the `Node` read model, and can remove the checks then.

bwaidelich commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php:
```diff
@@ -68,9 +68,16 @@ public function __construct(
         public readonly Timestamps $timestamps,
     ) {
+        if ($this->classification->isTethered() && $this->nodeName === null) {
+            throw new \InvalidArgumentException('The NodeName must be set if the Node is tethered.', 1695118377232);
```
this ain't a unix timestamp 👁️ 
```suggestion
            throw new \InvalidArgumentException('The NodeName must be set if the Node is tethered.', 1695118377);
```

mhsdesign replied:
lol then my tool is wrong - hold on 😱

mhsdesign replied:
<img width="481" alt="image" src="https://github.com/neos/neos-development-collection/assets/85400359/5ff7687a-3b6c-45b2-b5f0-8fa4813c4c82">

hmmm this indeed adds three more digits 😭 

it has to be like this: https://www.in2code.de/en/recent/phpstorm-insert-unix-timestamp-in-php/
thank you for pointing this out 😅

bwaidelich commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php:
```diff
@@ -68,9 +68,16 @@ public function __construct(
          * @return PropertyCollectionInterface Property values, indexed by their name
          */
         public readonly PropertyCollectionInterface $properties,
+        /**
```
```suggestion

        /**
```

mhsdesign replied:
hmm id say we leave it as is and move all the param docs into the main constructor comment like you did with the commands?

bwaidelich replied:
ah, sorry, I did't really check the surrounding code, just stumbled upon the missing empty line.

+1 for moving to `@param` annotations, otherwise the IDE won't show these

bwaidelich reviewed (commented):
just two mini tweaks, +1 apart from those

mhsdesign commented:
Hmm but now our ProjectionIntegrityViolationDetection fails:

https://github.com/neos/neos-development-collection/blob/d614486eab40085111d199a695fc3498716f4fe4/Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/ProjectionIntegrityViolationDetection/TetheredNodesAreNamed.feature#L42


```
001 Scenario: Create node variants of different type                             # Features/ProjectionIntegrityViolationDetection/TetheredNodesAreNamed.feature:41
      When the event NodeAggregateWithNodeWasCreated was published with payload: # Features/ProjectionIntegrityViolationDetection/TetheredNodesAreNamed.feature:42
        InvalidArgumentException: The NodeName must be set if the Node is tethered. in /home/runner/work/neos-development-collection/neos-development-collection/neos-development-distribution/Packages/Neos/Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php:79
        Stack trace:
        #0 /home/runner/work/neos-development-collection/neos-development-collection/neos-development-distribution/Packages/Neos/Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/NodeFactory.php(67): Neos\ContentRepository\Core\Projection\ContentGraph\Node->__construct()
        #1 /home/runner/work/neos-development-collection/neos-development-collection/neos-development-distribution/Packages/Neos/Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/NodeFactory.php(166): Neos\ContentGraph\DoctrineDbalAdapter\Domain\Repository\NodeFactory->mapNodeRowToNode()
        #2 /home/runner/work/neos-development-collection/neos-development-collection/neos-development-distribution/Packages/Neos/Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraph.php(245): Neos\ContentGraph\DoctrineDbalAdapter\Domain\Repository\NodeFactory->mapNodeRowsToNodeAggregate()
        #3 /home/runner/work/neos-development-collection/neos-development-collection/neos-development-distribution/Data/Temporary/Testing/SubContextBehat/Cache/Code/Flow_Object_Classes/Neos_Neos_Fusion_Cache_GraphProjectorCatchUpHookForCacheFlushing.php(176): Neos\ContentGraph\DoctrineDbalAdapter\Domain\Repository\ContentGraph->findNodeAggregateById()
        #4 /home/runner/work/neos-development-collection/neos-development-collection/neos-development-distribution/Packages/Neos/Neos.ContentRepository.Core/Classes/Projection/DelegatingCatchUpHook.php(44): Neos\Neos\Fusion\Cache\GraphProjectorCatchUpHookForCacheFlushing_Original->onAfterEvent()
        #5 /home/runner/work/neos-development-collection/neos-development-collection/neos-development-distribution/Packages/Neos/Neos.ContentRepository.Core/Classes/ContentRepository.php(183): Neos\ContentRepository\Core\Projection\DelegatingCatchUpHook->onAfterEvent()
        #6 /home/runner/work/neos-development-collection/neos-development-collection/neos-development-distribution/Packages/Libraries/neos/eventstore/src/CatchUp/CatchUp.php(148): Neos\ContentRepository\Core\ContentRepository->Neos\ContentRepository\Core\{closure}()
        #7 /home/runner/work/neos-development-collection/neos-development-collection/neos-development-distribution/Packages/Neos/Neos.ContentRepository.Core/Classes/ContentRepository.php(192): Neos\EventStore\CatchUp\CatchUp->run()
        #8 /home/runner/work/neos-development-collection/neos-development-collection/neos-development-distribution/Data/Temporary/Testing/SubContextBehat/Cache/Code/Flow_Object_Classes/Neos_ContentRepositoryRegistry_Factory_ProjectionCatchUpTrigger_CatchUpTriggerWithSynchronousOption.php(70): Neos\ContentRepository\Core\ContentRepository->catchUpProjection()
        #9 /home/runner/work/neos-development-collection/neos-development-collection/neos-development-distribution/Packages/Neos/Neos.ContentRepository.Core/Classes/EventStore/EventPersister.php(70): Neos\ContentRepositoryRegistry\Factory\ProjectionCatchUpTrigger\CatchUpTriggerWithSynchronousOption_Original->triggerCatchUp()
        #10 /home/runner/work/neos-development-collection/neos-development-collection/neos-development-distribution/Packages/Neos/Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/GenericCommandExecutionAndEventPublication.php(145): Neos\ContentRepository\Core\EventStore\EventPersister->publishEvents()
        #11 /home/runner/work/neos-development-collection/neos-development-collection/neos-development-distribution/Packages/Neos/Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/Features/NodeCreation.php(293): FeatureContext->publishEvent()
        #12 /home/runner/work/neos-development-collection/neos-development-collection/neos-development-distribution/Packages/Libraries/behat/behat/src/Behat/Testwork/Call/Handler/RuntimeCallHandler.php(110): FeatureContext->theEventNodeAggregateWithNodeWasCreatedWasPublishedToStreamWithPayload()
```


because of this (enabled) `GraphProjectorCatchUpHookForCacheFlushing` https://github.com/neos/neos-development-collection/blob/52be2f0c68fbb25fe64e575cc20a178fda5cca93/Neos.Neos/Classes/Fusion/Cache/GraphProjectorCatchUpHookForCacheFlushing.php#L176

A hack would be to disable it temporary? `private static bool $enabled = true;`

bwaidelich commented:
> because of this (enabled)

What is the root cause? Can't we fix that (e.g. adjust the test such that it does not create an invalid node)?

mhsdesign commented:
> What is the root cause? Can't we fix that (e.g. adjust the test such that it does not create an invalid node)?

as far as i understand is that we test that if a the database projection has an invalid node (somehow) we detect it.
For this to achieve we cannot use the normal commands but tell behat that those events are published (hacky! - but one of the few only ways, we should refrain from doing `Given the event` https://github.com/neos/neos-development-collection/issues/4336)
That results in some catchups being triggered that wont work because something is really wrong.

mhsdesign commented:
yes after adjusting the catch up the tests pass.

The neos.neos catchup is registers itself here into the default preset https://github.com/neos/neos-development-collection/blob/cc4a0a68657b7f72e95360cd4e99a1e93b21ab67/Neos.Neos/Configuration/Settings.ContentRepositoryRegistry.yaml#L18-L21

and since we use the content repository registry in `Neos.ContentRepository.BehavioralTests`

https://github.com/neos/neos-development-collection/blob/a177872daaf82a4b38bb5980224efca6bff4ad62/Neos.ContentRepository.BehavioralTests/Tests/Behavior/Bootstrap/FeatureContext.php#L164

we get everything registered for the default cr for free ;)

this is partially a good thing as we fully e2e test everything (at least that stuff generally works) but might also cause trouble like when doctrine is not setup but a projection needs it or it might slow down tests.

mhsdesign commented on Neos.ContentRepository.BehavioralTests/Tests/Behavior/Bootstrap/FeatureContext.php:
```diff
@@ -62,7 +63,23 @@ public function __construct()
         $this->setUpInterleavingLogger();
-        $this->contentRepositoryRegistry = $this->getObject(ContentRepositoryRegistry::class);
+
+        // todo hack we shouldn't rely on the registry here in the first place, but we do, and we have to exclude the `GraphProjectorCatchUpHookForCacheFlushingFactory`
+        // as in the ProjectionIntegrityViolationDetection/TetheredNodesAreNamed.feature:40 test we create invalid state to be recognized (via events)
+        // but the catchup wont let us create invalid state
+        // see https://github.com/neos/neos-development-collection/pull/4469#issuecomment-1807178264
+        // this _would_ be easily fixable by just hard coding all the information in here, but then we wont test or own default preset which is quite good that we do.
+        $contentRepositoryRegistrySettings = $this->getObject(ConfigurationManager::class)
+            ->getConfiguration(
+                ConfigurationManager::CONFIGURATION_TYPE_SETTINGS,
+                'Neos.ContentRepositoryRegistry'
+            );
+        unset($contentRepositoryRegistrySettings['presets']['default']['projections']['Neos.ContentRepository:ContentGraph']['catchUpHooks']['Neos.Neos:FlushContentCache']);
+
+        $this->contentRepositoryRegistry = new ContentRepositoryRegistry(
+            $contentRepositoryRegistrySettings,
+            $this->getObject(ObjectManagerInterface::class)
+        );
```
im not sure if youll like this hack :/

mhsdesign committed:
TASK: Manipulate content graph manually for TetheredNodesAreNamed test

otherwise it's basically impossible now to reproduce the state, as publishing an event will lead to the catchup's to run, which might instantiate a node:

the neos.neos `GraphProjectorCatchUpHookForCacheFlushingFactor` catchUpHook would fail for example:

> When the event NodeAggregateWithNodeWasCreated was published with payload:
> # Features/ProjectionIntegrityViolationDetection/TetheredNodesAreNamed.feature:41
> InvalidArgumentException: The NodeName must be set if the Node is tethered.

mhsdesign committed:
TASK: Add @var types to \TestSuite\GenericCommandExecutionAndEventPublication::publishEvent

... where its impossible to infer types

mhsdesign commented:
Thanks for all your input today. Inspired from some other tests in the doctrine dbal adapter:

To modifying the db directly to provoke this invalid state, i moved the test over there and did the same ;)

see https://github.com/neos/neos-development-collection/pull/4469/commits/cd8408c9b40ba283dcf44f01aa743857a7aa6e4f

hopefully this can now be finally merged :D Didnt expect this to blow up that much :P

bwaidelich reviewed (approved):
Nice one, thanks!

mhsdesign commented:
Thank you all for the extensive discussion last Friday (@kitsunet @bwaidelich @nezaniel) and sorry for hijacking that weekly partly for this non critical thing :D Im happy that this is older task of mine is now resolved ;)
